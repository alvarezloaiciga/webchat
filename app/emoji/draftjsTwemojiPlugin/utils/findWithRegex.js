const findWithRegex = (regex, contentBlock, callback) => {
  // Get the text from the contentBlock
  const text = contentBlock.getText();
  let matchArr;
  let start; // eslint-disable-line
  // Go through all matches in the text and return the indices to the callback
  // eslint-disable-next-line no-cond-assign
  while ((matchArr = regex.exec(text)) !== null) {
    start = matchArr.index;
    callback(start, start + matchArr[0].length);
  }
};

export default findWithRegex;
