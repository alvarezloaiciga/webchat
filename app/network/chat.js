// @flow

import QUIQ from 'utils/quiq';
import fetch from 'isomorphic-fetch';
import type { Conversation } from 'types';

const { HOST, CONTACT_POINT } = QUIQ;

const parseResponse = (response: Response): Promise<*> => {
  if (response.status && response.status >= 300) {
    return response.json()
      .then((res) => Promise.reject(res))
      .catch((err) => Promise.reject(err));
  }

  return response.json();
};

export const joinChat = () => {
  fetch(`${ HOST }/api/v1/messaging/chat/${ CONTACT_POINT }/join`, {
    mode: 'cors',
    credentials: 'include',
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
  });
};

export const leaveChat = () => {
  fetch(`${ HOST }/api/v1/messaging/chat/${ CONTACT_POINT }/leave`, {
    mode: 'cors',
    credentials: 'include',
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
  });
};

export const ping = () => {
  fetch(`${ HOST }/api/v1/webchat/endpoints/${ CONTACT_POINT }/socket-ping`, {
    mode: 'cors',
    method: 'POST',
    credentials: 'include',
  });
};

export const addMessage = (body: string) => {
  fetch(`${ HOST }/api/v1/messaging/chat/${ CONTACT_POINT }/send-message`, {
    mode: 'cors',
    credentials: 'include',
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({body}),
  });
};

export const fetchWebsocketInfo = (): Promise<{ url: string }> => (
  fetch(`${ HOST }/api/v1/messaging/chat/${ CONTACT_POINT }/socket-info`, {
    mode: 'cors',
    credentials: 'include',
  }).then(parseResponse)
);

export const fetchConversation = (): Promise<Conversation> => (
  fetch(`${ HOST }/api/v1/messaging/chat/${ CONTACT_POINT }`, {
    mode: 'cors',
    credentials: 'include',
  }).then(parseResponse)
);
