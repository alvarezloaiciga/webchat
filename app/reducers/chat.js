// @flow
import {inStandaloneMode} from 'Common/Utils';
import {createStore} from 'redux';
import {ChatInitializedState} from 'Common/Constants';
import quiqOptions from 'utils/quiq';
import type {ChatState, Action, ChatInitializedStateType, Message} from 'types';

type ChatAction = {
  chatContainerHidden?: boolean,
  agentsAvailable?: boolean,
  initializedState?: ChatInitializedStateType,
  transcript?: Array<Message>,
  agentTyping?: boolean,
};

export const initialState = {
  chatContainerHidden: true,
  agentsAvailable: false,
  initializedState: ChatInitializedState.UNINITIALIZED,
  transcript: [],
  agentTyping: false,
  welcomeFormRegistered: !quiqOptions.welcomeForm,
};

const chat = (state: ChatState, action: Action & ChatAction) => {
  switch (action.type) {
    case 'CHAT_CONTAINER_HIDDEN':
      return Object.assign({}, state, {
        chatContainerHidden: inStandaloneMode() ? false : action.chatContainerHidden,
      });
    case 'AGENTS_AVAILABLE':
      return Object.assign({}, state, {
        agentsAvailable: action.agentsAvailable,
      });
    case 'CHAT_INITIALIZED_STATE': {
      if (state.initializedState === ChatInitializedState.BURNED) {
        // One does not simply become unburned.
        return state;
      }

      return Object.assign({}, state, {initializedState: action.initializedState});
    }
    case 'UPDATE_TRANSCRIPT':
      return Object.assign({}, state, {transcript: action.transcript || []});
    case 'AGENT_TYPING':
      return Object.assign({}, state, {agentTyping: action.agentTyping});
    case 'WELCOME_FORM_REGISTERED':
      return Object.assign({}, state, {welcomeFormRegistered: true});
    case 'NEW_WEBCHAT_SESSION':
      // Reset state to initial state.
      // We keep the visibility state from before the new session, and we set initialized state to LOADING (while socket reconnects)
      return Object.assign(
        {},
        {...initialState},
        {
          chatContainerHidden: state.chatContainerHidden,
          agentsAvailable: state.agentsAvailable,
          initializedState: ChatInitializedState.LOADING,
        },
      );
    default:
      return state;
  }
};

export default chat;

// Selectors
export const getChatContainerHidden = (state: ChatState): boolean => {
  return state.chatContainerHidden;
};

export const getAgentsAvailable = (state: ChatState): boolean => {
  return state.agentsAvailable;
};