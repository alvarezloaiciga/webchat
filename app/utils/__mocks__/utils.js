export const getBrowserName = () => 'Chrome';
export const getMajor = () => 60;
export const getDeviceType = () => 'PC';
export const getOSName = () => 'Windows';
export const getUAInfo = () => ({
  browser: {
    name: 'Chrome',
    major: 60,
  },
  os: {
    name: 'Mac OS',
    version: 10,
  },
  device: {
    model: 'iPad',
    vendor: 'Apple',
    type: 'tablet',
  },
  cpu: {
    architecture: 'amd64',
  },
  engine: {
    name: 'WebKit',
    version: 537.36,
  },
  navigator:
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.81 Safari/537.36',
});
export const compatibilityMode = () => false;
export const isIE9 = () => false;
export const isIE10 = () => false;
export const nonCompatibleBrowser = () => false;
export const supportsFlexbox = () => true;
export const supportsSVG = () => true;
export const displayError = (error: string) => error;
export const isIEorSafari = () => false;
