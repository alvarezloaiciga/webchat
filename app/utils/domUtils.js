import ReactDOM, {render} from 'react-dom';
import React from 'react';
import Routes from 'Routes';
import {quiqContainerId} from 'Common/Constants';
import {AppContainer} from 'react-hot-loader';
import {IntlProvider} from 'react-intl';
import Redbox from 'redbox-react';
import {Provider} from 'react-redux';

export const appIsMounted = (): boolean => {
  const container = document.getElementById(quiqContainerId);
  if (!container) return false;
  return container.childElementCount > 0;
};

export const destructApp = () => {
  ReactDOM.unmountComponentAtNode(document.getElementById(quiqContainerId));
};

export const constructApp = store => {
  render(
    <Provider store={store}>
      <IntlProvider locale="en">
        <AppContainer errorReporter={Redbox}>
          <Routes />
        </AppContainer>
      </IntlProvider>
    </Provider>,
    document.getElementById(quiqContainerId),
  );

  if (module.hot) {
    module.hot.accept('Routes', () => {
      const NextApp = require('Routes').default; // eslint-disable-line global-require

      render(
        <Provider store={store}>
          <IntlProvider locale="en">
            <AppContainer errorReporter={Redbox}>
              <NextApp />
            </AppContainer>
          </IntlProvider>
        </Provider>,
        document.getElementById(quiqContainerId),
      );
    });
  }
};

export const removeElement = (e: HTMLElement) => {
  e.parentNode.removeChild(e);
};

export const reloadApp = store => {
  destructApp();
  constructApp(store);
};

export const setFavicon = (iconUrl: string, appleTouchUrl: ?string) => {
  // Remove all icons
  const icons = [
    ...Array.from(document.querySelectorAll("link[rel='icon']")),
    ...Array.from(document.querySelectorAll("link[rel='apple-touch-icon']")),
  ];
  icons.forEach(icon => {
    removeElement(icon);
  });

  // Build icon
  const icon = document.createElement('link');
  icon.rel = 'icon';
  icon.href = iconUrl;

  // Build apple touch icon
  const appleTouchIcon = document.createElement('link');
  appleTouchIcon.rel = 'apple-touch-icon';
  appleTouchIcon.href = appleTouchUrl || iconUrl;

  // NOTE HARD: The apple-touch-icon must come BEFORE the regular icon.
  // Chrome looks for the last meta tag with a rel containing "icon" to try and use as title bar icon.
  const head = document.getElementsByTagName('head')[0];
  head.appendChild(appleTouchIcon);
  head.appendChild(icon);
};

export const setApplicationName = (name: string) => {
  const appName = document.querySelector("meta[name='application-name']");
  if (appName) {
    removeElement(appName);
  }
  const newAppName = document.createElement('meta');
  newAppName.name = 'application-name';
  newAppName.content = name;
  document.getElementsByTagName('head')[0].appendChild(newAppName);
};

export const setBrowserThemeColor = (color: string) => {
  const colorTag = document.querySelector("meta[name='theme-color']");
  if (colorTag) {
    removeElement(colorTag);
  }
  const newColor = document.createElement('meta');
  newColor.name = 'theme-color';
  newColor.content = color;
  document.getElementsByTagName('head')[0].appendChild(newColor);
};
