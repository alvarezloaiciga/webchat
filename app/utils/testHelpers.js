// @flow
import type {Message, IntlObject, IntlMessage} from 'types';

export const getMockMessage = (
  // eslint-disable-line import/prefer-default-export
  id?: number = 0,
  overrides?: Message | {} = {},
) => {
  const message = {
    authorType: id % 2 === 0 ? 'Customer' : 'Agent',
    text: `text--${id}`,
    id: `id--${id}`,
    timestamp: id,
    type: 'Text',
  };

  return Object.assign(message, overrides);
};

/**
 * Stub out formatMessage so that it will return the defaultMessage with basic variable replacement
 */
const formatTestMessage = (msg: IntlMessage, values?: {[key: string]: string}) => {
  if (!values) {
    return msg.defaultMessage;
  }

  let returnValue = msg.defaultMessage;
  Object.keys(values).forEach(key => {
    if (values) {
      const value = values[key];
      returnValue = returnValue.replace(`{${key}}`, value);
    }
  });

  return returnValue;
};

export const TestIntlObject: IntlObject = {
  formatMessage: formatTestMessage,
  formatRelative: value => value.toString(),
  formatTime: timestamp => timestamp.toString(),
  formatDate: timestamp => timestamp.toString(),
};
