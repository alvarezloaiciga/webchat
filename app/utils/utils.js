// @flow
declare var Modernizr: Object;
import {UAParser} from 'ua-parser-js';
import type {BrowserNames, DeviceTypes, OSNames} from 'types';

const parser = new UAParser();

export const getBrowserName = (): BrowserNames => parser.getResult().browser.name;

export const getMajor = (): number => parseInt(parser.getResult().browser.major, 10);

export const getDeviceType = (): DeviceTypes => parser.getResult().device.type;

export const getOSName = (): OSNames => parser.getResult().os.name;

export const getUAInfo = () => parser.getResult();

export const compatibilityMode = () => {
  const compatList = [
    {
      name: 'IE',
      major: 10,
    },
  ];
  const {name, major} = parser.getResult().browser;
  return !!compatList.find(i => i.name === name && parseInt(major, 10) <= i.major);
};

export const isIE9 = () => getBrowserName() === 'IE' && getMajor() <= 9;
export const isIE10 = () => getBrowserName() === 'IE' && getMajor() === 10;

export const nonCompatibleBrowser = () => getBrowserName() === 'IE' && getMajor() < 9;
// Modernizr says IE10 doesn't support flexbox.
// It kind of does, at least for what we need it for... so go ahead and ignore Modernizr in that case
export const supportsFlexbox = () => isIE10() || Modernizr.flexbox;
