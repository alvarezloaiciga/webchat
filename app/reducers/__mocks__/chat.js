import {getMockConfiguration} from 'utils/testHelpers';

export const getConfiguration = () => getMockConfiguration();
export const getMessage = message => getMockConfiguration().messages[message].defaultMessage;
