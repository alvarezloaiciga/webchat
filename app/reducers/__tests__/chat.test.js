// @flow
jest.mock('Common/Utils');
jest.mock('Common/QuiqOptions');

import chat, {initialState} from '../chat';
import {getMockMessage} from 'utils/testHelpers';
import {inStandaloneMode} from 'Common/Utils';

describe('chat reducers', () => {
  afterEach(() => {
    (inStandaloneMode: any).mockReset();
  });

  describe('CHAT_CONTAINER_HIDDEN', () => {
    it('updates state with the new value', () => {
      expect(
        chat(initialState, {
          type: 'CHAT_CONTAINER_HIDDEN',
          chatContainerHidden: false,
        }).chatContainerHidden,
      ).toBe(false);
    });
  });

  describe('CHAT_LAUNCHER_HIDDEN', () => {
    it('updates state with the new value', () => {
      expect(
        chat(initialState, {
          type: 'CHAT_LAUNCHER_HIDDEN',
          chatLauncherHidden: false,
        }).chatLauncherHidden,
      ).toBe(false);
    });

    it('always sets chatLauncherHidden to true in standalone mode', () => {
      (inStandaloneMode: any).mockReturnValue(true);
      expect(
        chat(initialState, {
          type: 'CHAT_LAUNCHER_HIDDEN',
          chatLauncherHidden: false,
        }).chatLauncherHidden,
      ).toBe(true);
    });
  });

  describe('CHAT_INITIALIZED_STATE', () => {
    it('updates state with the new value', () => {
      expect(
        chat(initialState, {
          type: 'CHAT_INITIALIZED_STATE',
          initializedState: 'initialized',
        }).initializedState,
      ).toBe('initialized');
    });

    describe('when already burned', () => {
      it('does not update state', () => {
        expect(
          chat(Object.assign({}, initialState, {initializedState: 'burned'}), {
            type: 'CHAT_INITIALIZED_STATE',
            initializedState: 'initialized',
          }).initializedState,
        ).toBe('burned');
      });
    });
  });

  describe('AGENTS_AVAILABLE', () => {
    it('updates state with the new value', () => {
      expect(chat(initialState, {type: 'AGENTS_AVAILABLE', popped: true})).toMatchSnapshot();
      expect(chat(initialState, {type: 'AGENTS_AVAILABLE', popped: false})).toMatchSnapshot();
    });
  });

  describe('AGENT_TYPING', () => {
    it('updates state with the new value', () => {
      expect(chat(initialState, {type: 'AGENT_TYPING', agentTyping: true}).agentTyping).toBe(true);
    });
  });

  describe('WELCOME_FORM_REGISTERED', () => {
    it('updates state with the new value', () => {
      expect(chat(initialState, {type: 'WELCOME_FORM_REGISTERED'}).welcomeFormRegistered).toBe(
        true,
      );
    });
  });

  describe('NEW_WEBCHAT_SESSION', () => {
    const mock = getMockMessage();
    it('updates state with initial state merged with current visibility state and "initialized" set to "LOADING"', () => {
      const mutatedState = {
        welcomeFormRegistered: true,
        initializedState: 'initialized',
        chatContainerHidden: false,
        popped: false,
        chatLauncherHidden: false,
        transcript: {[mock.id]: mock},
        agentTyping: true,
        agentsAvailable: true,
        agentEndedConversation: false,
        platformEvents: [],
        muteSounds: false,
        messageFieldFocused: false,
        configuration: {
          enableChatEmailTranscript: false,
          enableChatFileAttachments: false,
          supportedAttachmentTypes: ['image/png,image/jpeg'],
          enableEmojis: false,
          playSoundOnNewMessage: false,
          flashNotificationOnNewMessage: false,
        },
        chatIsSpam: true,
      };

      expect(chat(mutatedState, {type: 'NEW_WEBCHAT_SESSION'})).toMatchSnapshot();
    });
  });
});
