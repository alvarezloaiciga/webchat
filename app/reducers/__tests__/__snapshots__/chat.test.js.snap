// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chat reducers AGENTS_AVAILABLE updates state with the new value 1`] = `
Object {
  "agentTyping": false,
  "agentsAvailable": undefined,
  "chatContainerHidden": true,
  "chatLauncherHidden": true,
  "configuration": Object {
    "enableChatEmailTranscript": false,
    "enableChatFileAttachments": false,
    "enableEmojis": false,
    "enableManualConvoStart": false,
    "flashNotificationOnNewMessage": false,
    "menuOptions": Object {
      "customItems": Array [],
      "offset": undefined,
    },
    "playSoundOnNewMessage": false,
    "registrationForm": undefined,
    "supportedAttachmentTypes": Array [
      "image/png,image/jpeg",
    ],
  },
  "initializedState": "uninitialized",
  "inputtingEmail": false,
  "isAgentAssigned": false,
  "messageFieldFocused": false,
  "muteSounds": false,
  "platformEvents": Object {},
  "transcript": Object {},
  "welcomeFormRegistered": false,
}
`;

exports[`chat reducers AGENTS_AVAILABLE updates state with the new value 2`] = `
Object {
  "agentTyping": false,
  "agentsAvailable": undefined,
  "chatContainerHidden": true,
  "chatLauncherHidden": true,
  "configuration": Object {
    "enableChatEmailTranscript": false,
    "enableChatFileAttachments": false,
    "enableEmojis": false,
    "enableManualConvoStart": false,
    "flashNotificationOnNewMessage": false,
    "menuOptions": Object {
      "customItems": Array [],
      "offset": undefined,
    },
    "playSoundOnNewMessage": false,
    "registrationForm": undefined,
    "supportedAttachmentTypes": Array [
      "image/png,image/jpeg",
    ],
  },
  "initializedState": "uninitialized",
  "inputtingEmail": false,
  "isAgentAssigned": false,
  "messageFieldFocused": false,
  "muteSounds": false,
  "platformEvents": Object {},
  "transcript": Object {},
  "welcomeFormRegistered": false,
}
`;

exports[`chat reducers NEW_WEBCHAT_SESSION updates state with initial state merged with current visibility state and "initialized" set to "LOADING" 1`] = `
Object {
  "agentTyping": false,
  "agentsAvailable": true,
  "chatContainerHidden": false,
  "chatLauncherHidden": false,
  "configuration": Object {
    "enableChatEmailTranscript": false,
    "enableChatFileAttachments": false,
    "enableEmojis": false,
    "enableManualConvoStart": false,
    "flashNotificationOnNewMessage": false,
    "menuOptions": Object {
      "customItems": Array [],
      "offset": undefined,
    },
    "playSoundOnNewMessage": false,
    "registrationForm": undefined,
    "supportedAttachmentTypes": Array [
      "image/png,image/jpeg",
    ],
  },
  "initializedState": "loading",
  "inputtingEmail": false,
  "isAgentAssigned": false,
  "messageFieldFocused": false,
  "muteSounds": false,
  "platformEvents": Object {},
  "transcript": Object {},
  "welcomeFormRegistered": false,
}
`;
